backend/src/graphql/resolver/util/sendTransactionsToDltConnector.ts

```diff
@@ -0,0 +1,42 @@
+import { IsNull } from '@dbTools/typeorm'
+import { DltTransaction } from '@entity/DltTransaction'
+
+import { DltConnectorClient } from '@/apis/DltConnectorClient'
+import { backendLogger as logger } from '@/server/logger'
+
+export async function sendTransactionsToDltConnector(): Promise<void> {
+  const dltConnector = DltConnectorClient.getInstance()
+  if (dltConnector) {
+    const dltTransactions = await DltTransaction.find({
+      where: { messageId: IsNull() },
+      relations: ['transaction'],
+      order: { createdAt: 'ASC', id: 'ASC' },
+    })
+    for (const dltTx of dltTransactions) {
```

Comment:
Bad Idea. When 2 or more transactions called shortly after another while the first call of sendTransactionsToDltConnector isn't finished yet, some transactions could be transmitted multiply time to the dlt-connector. The call of transmitTransaction need ca. 1 seconds!
Better give the transaction ids as parameter maybe as array for send/receive set.
NodeJs is Single Threaded but async. After a transmitTransaction is fired and still not returned (io-wait) NodeJs goes one with another stuff, like getting the next transaction.

Reply:
That's a good argument and have to be considered.
Please correct my statement, if i'm wrong: 
the send-request of a transaction to dlt-connector will insert the data in the iota-tangle and the insert will be confirmed with a messageId immediatly. 
This returned messageId will be send back in the response of the sendTransaction-request. 
The period of waiting for a iota-milestone is not part of this request-response-loop.

Reply:
It is not waiting on iota milestone, it is waiting on iota message id. 
For generating iota message id, a bit of mining (proof-of-work) is required.
So the call of the Iota message send function from @iota/client inside the binary lib (compiled from rust) launch x threads (x = cpu core count) to try out different nonce until the result reached the difficulty target set from iota network, put the transaction onto the tangle and return with the message id.

And even if it would return immediately, it would not mean that no time would pass. Since at least two network connections are involved, it takes at least a few ms. It may not be noticed for a long time, but the longer it is in operation, the higher the probability that it will happen. And to find this error then could be very difficult, because it is very hard to trigger.

Reply:
You're right to say, that sending a transaction multiple times must be avoided.
On my opinion to avoid this a synchronysing monitor, which lock multi processing of the same code, will solve this possibility very easy.
That's the reason why i split the `dlt_transactions-entry creation` from ´dlt_transactions-entry sending`. The dlt_transaction creation must on my opinion be part of the regular transaction-creations, which have to be finished inside the regular business processing and in time with the user-response . All other actions like sending to dlt - and in the next step the calculations of community-balance - and processing the dlt-response could be splitted and be processed in the background and outside the users-business-process.

Reply:
The monitor prevents concurrent execution, that is correct.
If a second transaction is to be sent within the execution of the first transaction, the execution of sendTransactionsToDltConnector is skipped. That means the second transaction will not be sent to the dlt connector for the time being. Only when sendTransactionsToDltConnector is called again during the next transaction. 
Then two transactions are sent shortly after each other. Because you sort the data request by date, the order is still correct, which is good. 

I can't see any critical error anymore so it's ok for me. Problems will only occur if I consider the rule that the Iota milestone timestamp must not be more than 2 minutes away from the createAt of the transaction. This is how it is handled by Hedera.

Reply:
Good hint about the max time from createdAt of transaction, but as you say it is handled in this way by Hedera. What about iota? Do we have to consider too or is on iota no limit?
I ask, because my vision is to send a balance_date with the transaction-data and this will be very older than the current time, because of the need to migrate the still existing transactions.

Reply:
Iota does not know our payload, so it does not matter for Iota. It is about our own validation logic. But we can also adapt this for the import of previous transactions. In my Gradido node I have done this so that the deviation is allowed until the first live transaction. So you can only import old transactions at the beginning, once the first live transaction has come in, no more.